packages/camera/ios/Classes/CameraPlugin.m

```diff
@@ -244,6 +233,59 @@ - (void)captureToFile:(NSString *)path result:(FlutterResult)result {
                                                            cameraPosition:_captureDevice.position]];
 }
 
+- (void)setCaptureSessionPreset:(NSString *)resolutionPreset {
+  int presetIndex;
+  if ([resolutionPreset isEqualToString:@"high"]) {
+    presetIndex = 0;
+  } else if ([resolutionPreset isEqualToString:@"medium"]) {
+    presetIndex = 2;
+  } else {
+    NSAssert([resolutionPreset isEqualToString:@"low"], @"Unknown resolution preset %@",
+             resolutionPreset);
+    presetIndex = 3;
+  }
+
+  switch (presetIndex) {
+    case 0:
+      if ([_captureSession canSetSessionPreset:AVCaptureSessionPreset3840x2160]) {
+        _captureSession.sessionPreset = AVCaptureSessionPreset3840x2160;
+        _previewSize = CGSizeMake(3840, 2160);
+        break;
+      }
+    case 1:
+      if ([_captureSession canSetSessionPreset:AVCaptureSessionPreset1920x1080]) {
+        _captureSession.sessionPreset = AVCaptureSessionPreset1920x1080;
+        _previewSize = CGSizeMake(1920, 1080);
+        break;
+      }
+    case 2:
+      if ([_captureSession canSetSessionPreset:AVCaptureSessionPreset1280x720]) {
+        _captureSession.sessionPreset = AVCaptureSessionPreset1280x720;
+        _previewSize = CGSizeMake(1280, 720);
+        break;
+      }
+    case 3:
+      if ([_captureSession canSetSessionPreset:AVCaptureSessionPreset640x480]) {
+        _captureSession.sessionPreset = AVCaptureSessionPreset640x480;
+        _previewSize = CGSizeMake(640, 480);
+        break;
+      }
+    case 4:
+      if ([_captureSession canSetSessionPreset:AVCaptureSessionPreset352x288]) {
+        _captureSession.sessionPreset = AVCaptureSessionPreset352x288;
+        _previewSize = CGSizeMake(352, 288);
+        break;
+      }
+    default: {
+      NSException *exception = [NSException
+          exceptionWithName:@"NoAvailableCaptureSessionException"
+                     reason:@"No capture session available for current capture session."
+                   userInfo:nil];
+      @throw exception;
+    }
+  }
```

Comment:
What happens if we go into any of these cases and `_captureSession canSetSessionPreset:` returns false?

Reply:
So [here](https://developer.apple.com/documentation/avfoundation/avcapturesessionpreset?language=objc) is the list of all capture session presets.

For history of the plugin and why I decided on this method to choose resolution preset: 
Originally we would used [AVCaptureSessionPresetHigh](https://developer.apple.com/documentation/avfoundation/avcapturesessionpresethigh?language=objc)/[AVCaptureSessionPresetMedium](https://developer.apple.com/documentation/avfoundation/avcapturesessionpresetmedium?language=objc)/[AVCaptureSessionPresetLow](https://developer.apple.com/documentation/avfoundation/avcapturesessionpresetlow?language=objc) for choosing the resolution. Which you can see starting at this [line](https://github.com/flutter/plugins/blob/eaa1388b11ac3ee6143ad782e573b9b9e314bc4b/packages/camera/ios/Classes/CameraPlugin.m#L103). I assume the camera plugin uses low/medium/high presets to copy iOS's way of handling resolution. 

However, there is no consistent way to get the correct `previewSize` before starting the camera. You can see more about this on [stackoverflow](https://stackoverflow.com/questions/7790402/knowing-resolution-of-avcapturesessions-session-presets). We would use this [method](https://github.com/flutter/plugins/blob/eaa1388b11ac3ee6143ad782e573b9b9e314bc4b/packages/camera/ios/Classes/CameraPlugin.m#L122) to get the resolution, but it would sometimes return the wrong `previewSize`. 

Since I couldn't find any way to get the correct resolution until after returning `previewSize`, I just chose to create a ladder with defined resolution presets. This ladder is just the subset of resolutions available on iOS. And **should** work with each device. It works for iphone 6 and iphone X. Other apps would just use a [chart](https://stackoverflow.com/questions/12240625/avcapturesession-settings-of-low-medium-and-high-what-res-do-they-each-represe), but I felt that would be overly complicated for fixing this bug.

Reply:
Another option would be to make `previewSize` an asynchronous getter which we update on the platform side.